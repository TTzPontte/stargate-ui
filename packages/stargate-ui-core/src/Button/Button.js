import React from 'react';
import { useStyles } from '@pontte/stargate-ui-styles';
import clsx from 'clsx';

const styles = ({ spacing }) => {
  const base = {
    display: 'inline-flex',
    padding: [spacing(), spacing()],
    border: 0,
    borderRadius: spacing(2),
    cursor: 'pointer',
    outline: 'none',
    '&:hover': {
      opacity: .5,
    },
  };

  return {
    base,
  };
};

const Button = ({ children }) => {
  const classes = useStyles(styles);
  const className = clsx(classes.base);

  return (
    <button type="button" className={className}>
      {children}
    </button>
  );
};

export default Button;
