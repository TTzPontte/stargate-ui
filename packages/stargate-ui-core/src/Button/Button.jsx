import React from 'react';
import { useStyles } from '@pontte/stargate-ui-styles';
import clsx from 'clsx';

import Factory from '../Factory';
import Typography from '../Typography';

const styles = (theme) => {
  /**
   * @todo improve dynamic properties
   */
  const {
    active,
    spacing,
    palette,
    border,
    radius,
    mode,
    resets,
  } = theme;
  const { setLightness } = palette;

  const getColor = ({ color }) => (
    color && palette[color][mode].color
  );

  const getTextColor = ({ color }) => (
    color && palette[color][mode].text
  );

  const button = {
    ...resets.button,
    display: 'inline-flex',
    border: [[...border, 'transparent']],
    borderRadius: radius(10),
    cursor: 'pointer',
    transition: [['all', '.4s', 'ease']],
    textDecoration: 'none',
    marginLeft: 0,
    marginRight: 0,
    width: ({ full }) => (
      full && '100%'
    ),
    pointerEvents: ({ disabled }) => (
      disabled && 'none'
    ),
    '& > *': {
      flexBasis: 'auto',
      flexGrow: 1,
      maxWidth: 'auto'
    },
  };

  const buttonText = {
    [active()]: {
      backgroundColor: ({ disabled, ...props }) => (
        !disabled && setLightness(.94, getColor(props))
      ),
      borderColor: ({ disabled, ...props }) => (
        !disabled && setLightness(.96, getColor(props))
      ),
    },
  };

  const buttonContained = {
    color: ({ disabled, ...props }) => (
      disabled ? setLightness(.8, getColor(props)) : getTextColor(props)
    ),
    backgroundColor: ({ disabled, ...props }) => (
      disabled ? setLightness(.95, getColor(props)) : getColor(props)
    ),
    [active()]: {
      color: ({ disabled, ...props }) => (
        !disabled && getColor(props)
      ),
      backgroundColor: ({ disabled }) => (
        !disabled && 'transparent'
      ),
      borderColor: ({ disabled, ...props }) => (
        !disabled && getColor(props)
      ),
    },
  };

  const buttonOutlined = {
    borderColor: getColor,
    color: ({ disabled, ...props }) => (
      disabled ? setLightness(.85, getColor(props)) : getColor(props)
    ),
    borderColor: ({ disabled, ...props }) => (
      disabled ? setLightness(.85, getColor(props)) : getColor(props)
    ),
    [active()]: {
      color: ({ disabled, ...props }) => (
        disabled ? setLightness(.6, getColor(props)) : getTextColor(props)
      ),
      backgroundColor: ({ disabled, ...props }) => (
        !disabled && getColor(props)
      ),
      borderColor: ({ disabled, ...props }) => (
        !disabled && getColor(props)
      ),
    },
  };

  return {
    button,
    buttonText,
    buttonContained,
    buttonOutlined,
  };
};

const Button = (props) => {
  const {
    contained,
    outlined,
    disabled,
    large,
    full,
    element = 'button',
    color = 'default',
    onClick = () => {},
    children,
    className: inheritedClasses,
    ...factoryProps
  } = props;

  const [
    {
      buttonText: classButtonText,
      buttonContained: classButtonContained,
      buttonOutlined: classButtonOutlined,
      ...classes
    },
  ] = useStyles(styles, {
    color,
    disabled,
    large,
    full,
  });
  const className = clsx(
    Object.values(classes),
    {
      [classButtonText]: !contained && !outlined,
      [classButtonContained]: contained,
      [classButtonOutlined]: outlined,
    },
    inheritedClasses,
  );

  const typographyVariant = large ? 'body' : 'bodylower';

  const handleClick = ({ event }) => {
    if (disabled) {
      event.stopPropagation();
      return;
    }

    onClick();
  }

  const n = !large ? 1 : 1.2;
  const paddingY = (1.5 * n);
  const paddingX = (6 * n);

  return (
    <Factory
      paddingX={paddingX}
      paddingY={paddingY}
      marginX={.5}
      {...factoryProps}
      element={element}
      className={className}
      onClick={handleClick}
    >
      <Typography
        element="span"
        variant={typographyVariant}
        gutter={0}
        transform="uppercase"
        children={children}
      />
    </Factory>
  );
};

export default Button;
