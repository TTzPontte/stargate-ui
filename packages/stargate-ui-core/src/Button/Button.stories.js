import React, { Fragment } from 'react';

import Button from './Button';

const story = {
  title: 'Core/Button',
  component: Button,
  parameters: {},
};

const BasicUsage = () => (
  <Fragment>
    <Button>I'm the first component!</Button>
  </Fragment>
);

export { BasicUsage };
export default story;
