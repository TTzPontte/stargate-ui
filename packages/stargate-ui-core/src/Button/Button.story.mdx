import {
  Meta,
  Canvas,
  Status,
  Story,
  withAction,
} from '@pontte/stargate-ui-playground';

import Button from './Button';

export const ButtonStory = (props) => (
  <Button
    onClick={withAction('onClick')}
    children="Cool button"
    color="primary"
    contained
    {...props}
  />
);

<Meta title="Core/Button" component={Button} />

<Story
  name="Explore Properties"
  parameters={{
    docs: { disable: true },
    jest: ['Button.spec.jsx'],
  }}
>
  {ButtonStory.bind({})}
</Story>

## Button

<Status type="rc" />

### Contained Button

<Canvas>
  <Button contained>Default</Button>
  <Button contained color="primary">Primary</Button>
  <Button contained color="secondary">Secondary</Button>
  <Button contained disabled>Disable</Button>
</Canvas>

### Outlined Button

<Canvas>
  <Button outlined>Default</Button>
  <Button color="primary" outlined>Primary</Button>
  <Button color="secondary" outlined>Secondary</Button>
  <Button outlined disabled>Disable</Button>
</Canvas>

### Text Button

<Canvas>
  <Button>Default</Button>
  <Button color="primary">Primary</Button>
  <Button color="secondary">Secondary</Button>
  <Button disabled>Disable</Button>
</Canvas>

### Sizes

<Canvas>
  <Button color="primary" contained>Small</Button>
  <Button
    color="primary"
    contained
    large
  >Large</Button>
</Canvas>
