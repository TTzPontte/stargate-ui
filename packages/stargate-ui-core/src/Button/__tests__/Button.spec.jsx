import React from 'react';

import Button from '../Button';

describe('Button', () => {
  it('should render properly', () => {
    const { container } = render(
      <Button>
        Click here
      </Button>,
    );

    expect(container).toMatchSnapshot();
  });
});
