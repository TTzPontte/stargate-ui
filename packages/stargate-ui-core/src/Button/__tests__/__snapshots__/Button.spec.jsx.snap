// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render properly 1`] = `
<div>
  <button
    class="stargate-factory-0-2-9 stargate-factory-d0-0-2-10 stargate-button-0-2-1 stargate-button-d0-0-2-5 stargate-buttonText-0-2-2 stargate-buttonText-d1-0-2-6"
  >
    <span
      class="stargate-factory-0-2-13 stargate-factory-d0-0-2-14 stargate-typography-0-2-11 stargate-typography-d0-0-2-12"
    >
      Click here
    </span>
  </button>
</div>
`;
